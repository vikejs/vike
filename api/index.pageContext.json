{"_urlRewrite":null,"pageId":"/pages/api","routeParams":{}}