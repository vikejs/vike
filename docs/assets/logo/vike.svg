<svg width="291" height="291" viewBox="0 0 291 291" fill="none" xmlns="http://www.w3.org/2000/svg">
<g clip-path="url(#clip0_13_108)">
<mask id="mask0_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask0_13_108)">
<path d="M26.5662 81.9988V104.884L155.111 241.829L176.151 244.69L219.153 239.522L286.24 170.498V141.799L165.813 18.7919L118.658 1.25921L97.3421 4.21207L26.5662 81.9988Z" fill="#8F8F8F"/>
</g>
<mask id="mask1_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="-3" y="5" width="289" height="289">
<path d="M285.504 5.0369H-2.49829V293.039H285.504V5.0369Z" fill="white"/>
</mask>
<g mask="url(#mask1_13_108)">
<path d="M17.1085 265.01C6.28625 253.482 2.80023 244.983 9.14287 245.579C15.4783 246.181 28.8278 255.62 39.6501 267.147C50.4724 278.675 53.9584 287.175 47.6229 286.579C41.2803 285.976 27.9302 276.538 17.1085 265.01Z" fill="#A6A6A6"/>
</g>
<mask id="mask2_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="-3" y="5" width="289" height="289">
<path d="M285.504 5.0369H-2.49829V293.039H285.504V5.0369Z" fill="white"/>
</mask>
<g mask="url(#mask2_13_108)">
<path fill-rule="evenodd" clip-rule="evenodd" d="M53.3212 287.523C51.8934 289.794 49.2607 290.173 47.302 289.989C45.0979 289.779 42.6276 288.861 40.1763 287.642C37.6642 286.393 34.9092 284.695 32.0685 282.688C26.3861 278.671 20.1704 273.275 14.6119 267.354C9.1231 261.507 5.19488 256.131 3.52026 252.047C2.75574 250.182 2.05049 247.629 3.05896 245.354C3.45608 244.458 4.01129 243.793 4.65233 243.307L17.5002 229.2L66.1138 273.476L53.3212 287.523Z" fill="url(#paint0_linear_13_108)"/>
</g>
<mask id="mask3_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="-3" y="5" width="289" height="289">
<path d="M285.504 5.0369H-2.49829V293.039H285.504V5.0369Z" fill="white"/>
</mask>
<g mask="url(#mask3_13_108)">
<path d="M30.5336 250.274C19.7113 238.746 16.2253 230.247 22.5607 230.849C28.9034 231.452 42.2529 240.884 53.0746 252.411C63.8968 263.945 67.3829 272.446 61.0474 271.843C54.7048 271.24 41.3559 261.809 30.5336 250.274Z" stroke="url(#paint1_linear_13_108)" stroke-width="6.84965" stroke-linecap="round" stroke-linejoin="round"/>
</g>
<mask id="mask4_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask4_13_108)">
<path fill-rule="evenodd" clip-rule="evenodd" d="M65.4975 267.832C65.0799 268.671 64.1199 269.062 62.5814 268.917C56.2454 268.314 42.8959 258.876 32.0742 247.348C22.9302 237.615 19.0217 230.039 21.846 228.28L74.9493 169.682L118.498 209.348L65.4975 267.832Z" fill="url(#paint2_linear_13_108)"/>
</g>
<mask id="mask5_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask5_13_108)">
<path d="M85.4523 188.452C96.2746 199.985 109.623 209.417 115.96 210.02C122.302 210.623 118.816 202.123 107.994 190.588C97.1716 179.061 83.8221 169.629 77.4794 169.026C71.144 168.423 74.63 176.924 85.4523 188.452Z" fill="#A56C4A"/>
</g>
<mask id="mask6_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask6_13_108)">
<path fill-rule="evenodd" clip-rule="evenodd" d="M121.399 208.714C120.473 210.289 118.655 210.662 116.923 210.5C115.028 210.32 112.781 209.509 110.399 208.325C107.987 207.125 105.309 205.478 102.518 203.506C96.9352 199.559 90.8056 194.24 85.3209 188.398C79.8711 182.593 76.1187 177.401 74.5647 173.615C73.8243 171.81 73.3809 169.851 74.0817 168.268C74.3601 167.64 74.7603 167.171 75.2391 166.83L88.2154 152.575L134.307 194.534L121.399 208.714Z" fill="url(#paint3_linear_13_108)"/>
</g>
<mask id="mask7_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask7_13_108)">
<path d="M99.9965 172.491C110.818 184.019 124.168 193.45 130.51 194.053C136.846 194.656 133.36 186.156 122.537 174.621C111.716 163.094 98.3663 153.662 92.0236 153.06C85.6882 152.457 89.1742 160.957 99.9965 172.491Z" fill="#9E9E9E" stroke="#9E9E9E" stroke-width="3.4247" stroke-linecap="round" stroke-linejoin="round"/>
</g>
<mask id="mask8_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask8_13_108)">
<path d="M287.462 141.957V142.98L246.959 139.113L116.782 0.770209L118.13 -7.62939e-05L165.875 17.6874L287.462 141.957Z" fill="url(#paint4_linear_13_108)"/>
</g>
<mask id="mask9_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask9_13_108)">
<path d="M219.589 240.856L176.297 245.869L175.927 245.601L247.339 166.875H247.761L287.47 170.404V170.946L219.589 240.856Z" fill="#B6B6B6"/>
</g>
<mask id="mask10_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask10_13_108)">
<path d="M96.8345 2.89326L118.123 0.00363922L119.703 1.82285L95.6938 14.8327L96.8345 2.89326Z" fill="#E5E5E5"/>
</g>
<mask id="mask11_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask11_13_108)">
<path d="M25.2503 92.6248L25.2498 104.672L155.013 242.907L155.634 243.021V228.209L25.8594 91.966L25.2503 92.6248Z" fill="#8B8A8B"/>
</g>
<mask id="mask12_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask12_13_108)">
<path d="M96.8404 2.86596L25.2479 81.4605V94.1044L96.8464 15.6238L96.8404 2.86596Z" fill="#E2E1E1"/>
</g>
<mask id="mask13_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask13_13_108)">
<path d="M25.2498 92.5288V101.388L29.0458 96.6305L25.2498 92.5288Z" fill="#8B8A8B"/>
</g>
<mask id="mask14_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask14_13_108)">
<path d="M96.8405 2.89326L118.129 0.00363922L96.8405 13.9969V2.89326Z" fill="#E5E5E5"/>
</g>
<mask id="mask15_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask15_13_108)">
<path d="M98.1425 13.2407L96.9035 14.0193L25.245 92.5958L156.052 231.875L228.724 152.176L98.1425 13.2407Z" fill="url(#paint5_linear_13_108)"/>
</g>
<mask id="mask16_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask16_13_108)">
<path d="M165.151 102.279L136.007 81.5596C135.528 81.2236 134.89 81.2788 134.459 81.6964L107.636 107.896C107.006 108.519 107.095 109.608 107.808 110.08L116.157 115.587C116.945 116.108 116.959 117.327 116.185 117.889L104.301 126.58C103.499 127.163 103.554 128.457 104.403 128.93L109.862 131.998C110.712 132.471 110.767 133.759 109.966 134.348L91.2526 147.957C90.0814 148.807 90.8488 150.759 92.2186 150.423L93.1294 150.197L152.623 127.841C153.616 127.464 153.801 126.026 152.932 125.402L144.596 119.443C143.815 118.882 143.87 117.614 144.699 117.108L165.055 104.615C165.884 104.108 165.939 102.84 165.151 102.279Z" fill="url(#paint6_linear_13_108)"/>
</g>
<mask id="mask17_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask17_13_108)">
<path d="M176.299 245.866L155.011 242.972V230.646L155.26 230.389L176.641 245.582L176.299 245.866Z" fill="#939493"/>
</g>
<mask id="mask18_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask18_13_108)">
<path d="M247.886 138.237L226.598 152.23L96.8351 13.9955L118.123 0.00219727L247.886 138.237Z" fill="url(#paint7_linear_13_108)"/>
</g>
<mask id="mask19_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask19_13_108)">
<path d="M287.45 170.955L247.463 167.255V138.22L287.45 141.914V170.955Z" fill="#F2F2F2"/>
</g>
<mask id="mask20_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask20_13_108)">
<path d="M247.861 167.247L176.299 245.866L154.977 230.809L226.815 152.035L247.862 166.731L247.861 167.247Z" fill="#9B9B9B"/>
</g>
<mask id="mask21_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask21_13_108)">
<path d="M247.891 167.271V138.237L247.721 137.998L226.464 152.013L226.603 152.23L247.891 167.271Z" fill="#A7A6A7"/>
</g>
<mask id="mask22_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask22_13_108)">
<path opacity="0.5" fill-rule="evenodd" clip-rule="evenodd" d="M165.751 103.401C165.751 103.78 165.492 104.338 165.169 104.536L144.666 117.085C144.612 117.138 144.54 117.211 144.463 117.297C144.29 117.494 144.147 117.703 144.093 117.869C143.922 118.385 144.081 118.967 144.327 119.258L152.76 125.28L152.794 125.312C152.871 125.386 153.03 125.452 153.097 125.534C153.192 125.647 153.52 126.016 153.496 126.511C153.473 127.006 153.289 127.57 152.794 127.756L93.1445 150.163C92.7905 150.296 91.9493 150.585 91.6211 150.421C91.2929 150.256 90.8477 150.093 90.8477 149.764L152.024 126.511C152.007 126.484 151.984 126.454 151.957 126.422C151.929 126.388 151.901 126.358 151.881 126.336C151.881 126.336 143.434 120.068 143.263 119.95C143.091 119.831 142.79 119.436 142.677 118.864C142.649 118.48 142.637 117.905 142.79 117.44C142.933 117.007 143.224 116.63 143.434 116.392C143.546 116.265 143.65 116.161 143.727 116.087C143.765 116.05 143.797 116.021 143.82 116L143.849 115.975L143.857 115.967L143.862 115.963C143.928 115.906 144.003 115.864 144.08 115.837L165.336 102.703C165.658 102.505 165.751 103.023 165.751 103.401Z" fill="white"/>
</g>
<mask id="mask23_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask23_13_108)">
<path opacity="0.3" d="M248.06 166.515L227.216 151.887" stroke="white" stroke-width="1.36987" stroke-linecap="round"/>
</g>
<mask id="mask24_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="4" y="-4" width="289" height="290">
<path d="M292.533 -3.00148H4.53052V285.001H292.533V-3.00148Z" fill="white"/>
</mask>
<g mask="url(#mask24_13_108)">
<path d="M173.755 59.423L208.535 133.2L226.645 152.248L247.95 138.208L173.755 59.423Z" fill="url(#paint8_linear_13_108)"/>
</g>
<mask id="mask25_13_108" style="mask-type:luminance" maskUnits="userSpaceOnUse" x="-3" y="5" width="289" height="289">
<path d="M285.504 5.0369H-2.49829V293.039H285.504V5.0369Z" fill="white"/>
</mask>
<g mask="url(#mask25_13_108)">
<path fill-rule="evenodd" clip-rule="evenodd" d="M22.205 227.972C22.205 229.168 22.2958 229.413 22.626 230.641C22.8924 231.632 23.4222 233.142 24.4738 235.131C26.5776 239.109 30.7733 245.007 39.1343 252.485C47.502 259.969 53.9616 263.828 58.2901 265.815C60.4543 266.808 62.0875 267.334 63.1621 267.61C63.6997 267.748 64.0981 267.824 64.3537 267.865C64.7713 267.931 65.0899 267.902 65.4607 267.902C65.1889 268.355 64.4317 269.266 64.1113 269.6C63.7909 269.933 62.9191 269.691 62.5645 269.6C61.3663 269.292 59.7043 268.725 57.4331 267.683C52.8905 265.598 46.2616 261.616 37.7644 254.016C29.2599 246.41 24.8967 240.326 22.6575 236.091C21.5375 233.973 20.9498 232.32 20.6417 231.174C20.4876 230.601 20.4695 230.465 20.4238 230.152C20.3019 229.317 21.443 228.41 22.205 227.972Z" fill="#E4E4E4"/>
</g>
</g>
<defs>
<linearGradient id="paint0_linear_13_108" x1="7.73661" y1="240.473" x2="55.7681" y2="284.459" gradientUnits="userSpaceOnUse">
<stop stop-color="#C4C4C4"/>
<stop offset="0.25" stop-color="#F8F8F8"/>
<stop offset="0.46763" stop-color="#A6A6A6"/>
<stop offset="0.68" stop-color="#B8B8B8"/>
<stop offset="1" stop-color="#A9AAA9"/>
</linearGradient>
<linearGradient id="paint1_linear_13_108" x1="17.0798" y1="230.146" x2="64.0002" y2="275.477" gradientUnits="userSpaceOnUse">
<stop stop-color="#C4C4C4"/>
<stop offset="0.25754" stop-color="#F8F8F8"/>
<stop offset="0.46864" stop-color="#A6A6A6"/>
<stop offset="0.67922" stop-color="#B8B8B8"/>
<stop offset="1" stop-color="#A9AAA9"/>
</linearGradient>
<linearGradient id="paint2_linear_13_108" x1="53.3649" y1="198.636" x2="94.4592" y2="236.31" gradientUnits="userSpaceOnUse">
<stop stop-color="#A67333"/>
<stop offset="0.235" stop-color="#AE7C3F"/>
<stop offset="0.47" stop-color="#C6832F"/>
<stop offset="0.715" stop-color="#E3983D"/>
<stop offset="1" stop-color="#E79934"/>
</linearGradient>
<linearGradient id="paint3_linear_13_108" x1="73.735" y1="167.688" x2="114.283" y2="210.534" gradientUnits="userSpaceOnUse">
<stop stop-color="#D5D5D4"/>
<stop offset="0.49" stop-color="#7D7E7C"/>
<stop offset="1" stop-color="#838283"/>
</linearGradient>
<linearGradient id="paint4_linear_13_108" x1="152.684" y1="0.0036399" x2="279.392" y2="143.844" gradientUnits="userSpaceOnUse">
<stop stop-color="#AFAFAF"/>
<stop offset="1" stop-color="#BFBEBF"/>
</linearGradient>
<linearGradient id="paint5_linear_13_108" x1="104.737" y1="30.8262" x2="169.808" y2="212.334" gradientUnits="userSpaceOnUse">
<stop stop-color="#BABAB9"/>
<stop offset="1" stop-color="#8F8F8F"/>
</linearGradient>
<linearGradient id="paint6_linear_13_108" x1="162.956" y1="102.743" x2="87.6131" y2="133.565" gradientUnits="userSpaceOnUse">
<stop stop-color="#FBCC56"/>
<stop offset="1" stop-color="#FEAD01"/>
</linearGradient>
<linearGradient id="paint7_linear_13_108" x1="115.009" y1="0.00363977" x2="234.872" y2="136.991" gradientUnits="userSpaceOnUse">
<stop stop-color="#E4E3E4"/>
<stop offset="1" stop-color="#ADADAC"/>
</linearGradient>
<linearGradient id="paint8_linear_13_108" x1="210.908" y1="59.423" x2="210.908" y2="152.346" gradientUnits="userSpaceOnUse">
<stop stop-color="#C3C3C3"/>
<stop offset="1" stop-color="#AAAAA9"/>
</linearGradient>
<clipPath id="clip0_13_108">
<rect width="290.03" height="290.03" fill="white"/>
</clipPath>
</defs>
</svg>
