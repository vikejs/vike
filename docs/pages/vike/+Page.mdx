See [vite-plugin-ssr.com/vike](https://vite-plugin-ssr.com/vike).
