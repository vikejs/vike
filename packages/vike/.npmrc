# Throw an error at install-time if user has a Node.js version lower than package.json#engines.node
engine-strict=true
