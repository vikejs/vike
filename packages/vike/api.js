// Some tools still need this as of January 2025
export * from './dist/esm/node/api/index.js'
export { default } from './dist/esm/node/api/index.js'
