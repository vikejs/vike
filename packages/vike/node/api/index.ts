export { build_ as build }
export { dev_ as dev }
export { preview_ as preview }
export { prerender_ as prerender }

import type { build } from './build.js'
import type { dev } from './dev.js'
import type { preview } from './preview.js'
import type { prerender } from './prerender.js'
import { assertIsNotBrowser } from '../../utils/assertIsNotBrowser.js'
assertIsNotBrowser()

const build_: typeof build = async (...args) => (await import('./build.js')).build(...args)
const dev_: typeof dev = async (...args) => (await import('./dev.js')).dev(...args)
const preview_: typeof preview = async (...args) => (await import('./preview.js')).preview(...args)
const prerender_: typeof prerender = async (...args) => (await import('./prerender.js')).prerender(...args)
