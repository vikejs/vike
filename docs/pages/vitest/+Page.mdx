import { Link } from '@brillout/docpress'

When running Vitest consider either:
 - Using Vike's JavaScript API, see <Link href="#with-vike" />.
 - Removing Vike's Vite plugin, see <Link href="#without-vike" />.

## With Vike

If you want to use Vitest to test your Vike app, then you can use <Link href="/api">Vike's JavaScript API</Link>:

```ts
// dev.spec.ts

import { afterAll, beforeAll } from 'vitest'
import { testApp, viteConfig } from './testApp'
import { dev } from 'vike/api'

let viteServer: Awaited<ReturnType<typeof dev>>['viteServer']

beforeAll(async () => {
  const { viteServer } = await dev({ viteConfig })
  await viteServer.listen()
  await sleep(10) // avoid race condition of server not actually being ready
}, 10 * 1000)

afterAll(async () => {
  await viteServer.close()
})

testApp()

function sleep(milliseconds: number): Promise<void> {
  return new Promise((r) => setTimeout(r, milliseconds))
}
```

```ts
// preview.spec.ts

import { afterAll, beforeAll } from 'vitest'
import { testApp, viteConfig } from './testApp'
import { build, preview } from 'vike/api'

let viteServer: Awaited<ReturnType<typeof preview>>['viteServer']

beforeAll(async () => {
  await build({ viteConfig })
  const { viteServer } = await preview({ viteConfig })
  // If you want to replicate Vite's startup log:
  viteServer!.printUrls()
  // For the preview server, no need to call server.listen()
}, 40 * 1000)

afterAll(async () => {
  await viteServer!.close()
})

testApp()
```

```ts
// testApp.ts

export { testApp }
export { viteConfig }

import { expect, describe, it } from 'vitest'

const viteConfig = {
  logLevel: 'warn' as const,
  root: __dirname,
  configFile: __dirname + '/vite.config.js'
}
const urlBase = 'http://localhost:3000'

function testApp() {
  describe('Vitest', () => {
    it('run Vitest with Vike', { timeout: 10 * 1000 }, async () => {
      {
        const html = await fetchHtml('/')
        expect(html).toContain('<h1>Welcome</h1>')
        expect(html).toContain('<li>Rendered to HTML.</li>')
      }
      {
        const html = await fetchHtml('/about')
        expect(html).toContain('<h1>About</h1>')
        expect(html).toContain('<p>Example of using Vike.</p>')
      }
    })
  })
}

async function fetchHtml(urlPathname: string) {
  const ret = await fetch(urlBase + urlPathname)
  const html = await ret.text()
  return html
}
```

> Feel free to [reach out on GitHub Discussions](https://github.com/vikejs/vike/discussions) if you want help setting up Vitest with Vike differently.


## Without Vike

If you want to test JavaScript code without testing your Vike app, then remove <Link href="/vite-plugin">Vike's Vite plugin</Link> when running Vitest:

```ts
// vite.config.ts

import type { UserConfig } from 'vite'
import vike from 'vike/plugin'

export default {
  plugin: [
    // Always add Vike's Vite plugin// [!code --]
    vike(),// [!code --]
    // Don't add Vike's Vite plugin when running Vitest// [!code ++]
    !process.env.VITEST && vike(),// [!code ++]
    // ...
  ]
} satisfies UserConfig
```

So that you can share your `vite.config.js` settings between Vitest and your Vike app.
