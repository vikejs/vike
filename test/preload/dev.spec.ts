import { beforeAll } from 'vitest'
import { dev } from 'vike/api'
import { testRun } from './testRun'

beforeAll(async () => {
  await devApp()
}, 10 * 1000)

testRun(true)

async function devApp() {
  await dev({
    viteConfig: {
      root: __dirname,
      server: {
        /* Doesn't seem to work
        hmr: false
        */
        hmr: {
          // Avoid:
          // ```
          // Error: listen EADDRINUSE: address already in use :::24678
          // ```
          port: 11323
        }
      }
    }
  })
}
