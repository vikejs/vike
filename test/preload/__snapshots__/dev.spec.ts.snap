// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preload tags > Preload Default 1`] = `
[
  {
    "assetType": "script",
    "earlyHintLink": "</@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js>; rel=modulepreload; as=script",
    "isEntry": true,
    "mediaType": "text/javascript",
    "src": "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js",
  },
  {
    "assetType": "style",
    "earlyHintLink": "</renderer/Layout.css?direct>; rel=preload; as=style",
    "isEntry": true,
    "mediaType": "text/css",
    "src": "/renderer/Layout.css?direct",
  },
]
`;

exports[`preload tags > Preload Default 2`] = `
"<!DOCTYPE html>
    <html>
      <head>
        <link rel="stylesheet" type="text/css" href="/renderer/Layout.css?direct">
      </head>
      <body>
        <div id="root"><div style="display:flex;max-width:900px;margin:auto"><div style="padding:20px;padding-top:20px;flex-shrink:0;display:flex;flex-direction:column;align-items:center;line-height:1.8em"><div style="margin-top:20px;margin-bottom:10px"><a href="/"><img src="/renderer/logo.svg" height="64" width="64"/></a></div><a class="navitem" href="/">Preload Default</a><a class="navitem" href="/preload-disabled">Preload Disabled</a><a class="navitem" href="/preload-images">Preload Images</a></div><div style="padding:20px;padding-bottom:50px;border-left:2px solid #eee;min-height:100vh"><h1>Default</h1><p>This page showcases the default preloading strategy: in production, both the image and the font are preloaded.</p></div></div></div>
        <script id="vike_pageContext" type="application/json">{"_urlRewrite":null,"pageId":"/pages/index","routeParams":{}}</script>
        <script type="module" async>
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
import "/@vite/client";
import "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js";
</script>
      </body>
    </html>"
`;

exports[`preload tags > Preload Disabled 1`] = `
[
  {
    "assetType": "script",
    "earlyHintLink": "</@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js>; rel=modulepreload; as=script",
    "isEntry": true,
    "mediaType": "text/javascript",
    "src": "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js",
  },
  {
    "assetType": "style",
    "earlyHintLink": "</renderer/Layout.css?direct>; rel=preload; as=style",
    "isEntry": true,
    "mediaType": "text/css",
    "src": "/renderer/Layout.css?direct",
  },
]
`;

exports[`preload tags > Preload Disabled 2`] = `
"<!DOCTYPE html>
    <html>
      <head>
        <link rel="stylesheet" type="text/css" href="/renderer/Layout.css?direct">
      </head>
      <body>
        <div id="root"><div style="display:flex;max-width:900px;margin:auto"><div style="padding:20px;padding-top:20px;flex-shrink:0;display:flex;flex-direction:column;align-items:center;line-height:1.8em"><div style="margin-top:20px;margin-bottom:10px"><a href="/"><img src="/renderer/logo.svg" height="64" width="64"/></a></div><a class="navitem" href="/">Preload Default</a><a class="navitem" href="/preload-disabled">Preload Disabled</a><a class="navitem" href="/preload-images">Preload Images</a></div><div style="padding:20px;padding-bottom:50px;border-left:2px solid #eee;min-height:100vh"><h1>Preload Disabled</h1><p>This page showcases completely disabled preloading: the image nor the font are preloaded.</p></div></div></div>
        <script id="vike_pageContext" type="application/json">{"_urlRewrite":null,"pageId":"/pages/preload-disabled","routeParams":{}}</script>
        <script type="module" async>
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
import "/@vite/client";
import "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js";
</script>
      </body>
    </html>"
`;

exports[`preload tags > Preload Eager 1`] = `
[
  {
    "assetType": "script",
    "earlyHintLink": "</@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js>; rel=modulepreload; as=script",
    "isEntry": true,
    "mediaType": "text/javascript",
    "src": "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js",
  },
  {
    "assetType": "style",
    "earlyHintLink": "</renderer/Layout.css?direct>; rel=preload; as=style",
    "isEntry": true,
    "mediaType": "text/css",
    "src": "/renderer/Layout.css?direct",
  },
]
`;

exports[`preload tags > Preload Eager 2`] = `
"<!DOCTYPE html>
    <html>
      <head>
        <link rel="stylesheet" type="text/css" href="/renderer/Layout.css?direct">
      </head>
      <body>
        <div id="root"><div style="display:flex;max-width:900px;margin:auto"><div style="padding:20px;padding-top:20px;flex-shrink:0;display:flex;flex-direction:column;align-items:center;line-height:1.8em"><div style="margin-top:20px;margin-bottom:10px"><a href="/"><img src="/renderer/logo.svg" height="64" width="64"/></a></div><a class="navitem" href="/">Preload Default</a><a class="navitem" href="/preload-disabled">Preload Disabled</a><a class="navitem" href="/preload-images">Preload Images</a></div><div style="padding:20px;padding-bottom:50px;border-left:2px solid #eee;min-height:100vh"><h1>Eager</h1><p>This page showcases eager preloading (non-JavaScript assets are preloaded ASAP).</p></div></div></div>
        <script id="vike_pageContext" type="application/json">{"_urlRewrite":null,"pageId":"/pages/preload-eager","routeParams":{}}</script>
        <script type="module" async>
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
import "/@vite/client";
import "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js";
</script>
      </body>
    </html>"
`;

exports[`preload tags > Preload Images 1`] = `
[
  {
    "assetType": "script",
    "earlyHintLink": "</@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js>; rel=modulepreload; as=script",
    "isEntry": true,
    "mediaType": "text/javascript",
    "src": "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js",
  },
  {
    "assetType": "style",
    "earlyHintLink": "</renderer/Layout.css?direct>; rel=preload; as=style",
    "isEntry": true,
    "mediaType": "text/css",
    "src": "/renderer/Layout.css?direct",
  },
]
`;

exports[`preload tags > Preload Images 2`] = `
"<!DOCTYPE html>
    <html>
      <head>
        <link rel="stylesheet" type="text/css" href="/renderer/Layout.css?direct">
      </head>
      <body>
        <div id="root"><div style="display:flex;max-width:900px;margin:auto"><div style="padding:20px;padding-top:20px;flex-shrink:0;display:flex;flex-direction:column;align-items:center;line-height:1.8em"><div style="margin-top:20px;margin-bottom:10px"><a href="/"><img src="/renderer/logo.svg" height="64" width="64"/></a></div><a class="navitem" href="/">Preload Default</a><a class="navitem" href="/preload-disabled">Preload Disabled</a><a class="navitem" href="/preload-images">Preload Images</a></div><div style="padding:20px;padding-bottom:50px;border-left:2px solid #eee;min-height:100vh"><h1>Preload Images</h1><p>This page showcases a custom strategy of preloading images.</p></div></div></div>
        <script id="vike_pageContext" type="application/json">{"_urlRewrite":null,"pageId":"/pages/preload-images","routeParams":{}}</script>
        <script type="module" async>
import RefreshRuntime from "/@react-refresh"
RefreshRuntime.injectIntoGlobalHook(window)
window.$RefreshReg$ = () => {}
window.$RefreshSig$ = () => (type) => type
window.__vite_plugin_react_preamble_installed__ = true
import "/@vite/client";
import "/@fs/$ROOT/vike/dist/esm/client/server-routing-runtime/entry.js";
</script>
      </body>
    </html>"
`;
